builtin/providers/vsphere/resource_vsphere_virtual_machine.go

```diff
@@ -0,0 +1,1062 @@
+							Type:     schema.TypeInt,
+							Optional: true,
+							ForceNew: true,
+						},
+					},
+				},
+			},
+
+			"boot_delay": &schema.Schema{
+				Type:     schema.TypeInt,
+				Optional: true,
+				ForceNew: true,
+			},
+		},
+	}
+}
+
+func resourceVSphereVirtualMachineCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*govmomi.Client)
+
+	vm := virtualMachine{
+		name:     d.Get("name").(string),
+		vcpu:     d.Get("vcpu").(int),
+		memoryMb: int64(d.Get("memory").(int)),
+	}
+
+	if v, ok := d.GetOk("datacenter"); ok {
+		vm.datacenter = v.(string)
+	}
+
+	if v, ok := d.GetOk("cluster"); ok {
+		vm.cluster = v.(string)
+	}
+
+	if v, ok := d.GetOk("resource_pool"); ok {
+		vm.resourcePool = v.(string)
+	}
+
+	if v, ok := d.GetOk("gateway"); ok {
+		vm.gateway = v.(string)
+	}
+
+	if v, ok := d.GetOk("domain"); ok {
+		vm.domain = v.(string)
+	}
+
+	if v, ok := d.GetOk("time_zone"); ok {
+		vm.timeZone = v.(string)
+	}
+
+	dns_suffix := d.Get("dns_suffix.#").(int)
+	if dns_suffix > 0 {
+		vm.dnsSuffixes = make([]string, 0, dns_suffix)
+		for i := 0; i < dns_suffix; i++ {
+			s := fmt.Sprintf("dns_suffix.%d", i)
+			vm.dnsSuffixes = append(vm.dnsSuffixes, d.Get(s).(string))
+		}
+	} else {
+		vm.dnsSuffixes = DefaultDNSSuffixes
+	}
+
+	dns_server := d.Get("dns_server.#").(int)
+	if dns_server > 0 {
+		vm.dnsServers = make([]string, 0, dns_server)
+		for i := 0; i < dns_server; i++ {
+			s := fmt.Sprintf("dns_server.%d", i)
+			vm.dnsServers = append(vm.dnsServers, d.Get(s).(string))
+		}
+	} else {
+		vm.dnsServers = DefaultDNSServers
+	}
+
+	networksCount := d.Get("network_interface.#").(int)
+	networks := make([]networkInterface, networksCount)
+	for i := 0; i < networksCount; i++ {
+		prefix := fmt.Sprintf("network_interface.%d", i)
+		networks[i].label = d.Get(prefix + ".label").(string)
+		if v, ok := d.GetOk(prefix + ".ip_address"); ok {
+			networks[i].ipAddress = v.(string)
+		}
+		if v, ok := d.GetOk(prefix + ".subnet_mask"); ok {
+			networks[i].subnetMask = v.(string)
+		}
+	}
+	vm.networkInterfaces = networks
+	log.Printf("[DEBUG] network_interface init: %v", networks)
+
+	diskCount := d.Get("disk.#").(int)
+	disks := make([]hardDisk, diskCount)
+	for i := 0; i < diskCount; i++ {
+		prefix := fmt.Sprintf("disk.%d", i)
+		if i == 0 {
+			if v, ok := d.GetOk(prefix + ".template"); ok {
+				vm.template = v.(string)
+			} else {
+				if v, ok := d.GetOk(prefix + ".size"); ok {
+					disks[i].size = int64(v.(int))
+				} else {
+					return fmt.Errorf("If template argument is not specified, size argument is required.")
+				}
+			}
+			if v, ok := d.GetOk(prefix + ".datastore"); ok {
+				vm.datastore = v.(string)
+			}
+		} else {
+			if v, ok := d.GetOk(prefix + ".size"); ok {
+				disks[i].size = int64(v.(int))
+			} else {
+				return fmt.Errorf("Size argument is required.")
+			}
+		}
+		if v, ok := d.GetOk(prefix + ".iops"); ok {
+			disks[i].iops = int64(v.(int))
+		}
+	}
```

Comment:
Similar story here. Rather than interacting the internal layout of the value, do a single `d.Get("disk")` which will yield a `[]interface{}` whose elements are `map[string]interface`. Examples of code in this style can be found in the [aws_instance resource](https://github.com/hashicorp/terraform/blob/master/builtin/providers/aws/resource_aws_instance.go).